depthai/plugin.cpp

```diff
@@ -0,0 +1,344 @@
+
+#include <cstdio>
+#include <iostream>
+
+#include "opencv2/opencv.hpp"
+
+// Inludes common necessary includes for development using depthai library
+#include "depthai/depthai.hpp"
+
+// ILLIXR includes
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+
+using namespace ILLIXR;
+
+class depthai : public plugin {
+public:
+    depthai(std::string name_, phonebook* pb_)
+        : plugin{name_, pb_}
+        , sb{pb->lookup_impl<switchboard>()}
+        , _m_imu_cam{sb->get_writer<imu_cam_type>("imu_cam")}
+        , _m_rgb_depth{sb->get_writer<rgb_depth_type>("rgb_depth")}
+        //Initialize DepthAI pipeline and device 
+        , p{createCameraPipeline()}
+        , d{p}
+        { 
+            d.startPipeline();
+            #ifndef NDEBUG
+                std::cout << "Depthai pipeline started" << std::endl;
+            #endif
+            color = d.getOutputQueue("preview", 1, false);
+            depthQueue = d.getOutputQueue("depth", 1, false);
+            rectifLeftQueue = d.getOutputQueue("rectified_left", 1, false);
+            rectifRightQueue = d.getOutputQueue("rectified_right", 1, false);
+            imuQueue = d.getOutputQueue("imu", 1, false);
+            std::function<void(void)> imu_callback = [&](){callback();};
+            imuQueue->addCallback(imu_callback);
+        }
+
+    void callback(){  
+        std::lock_guard<std::mutex> lock(mutex);
+        //Check for available data
+        bool color_go = color->has<dai::ImgFrame>();
+        bool depth_go = depthQueue->has<dai::ImgFrame>();
+        bool rectifL_go = rectifLeftQueue->has<dai::ImgFrame>();
+        bool rectifR_go = rectifRightQueue->has<dai::ImgFrame>();
+        bool imuPacket_go = imuQueue->has<dai::IMUData>();
+        
+        #ifndef NDEBUG
+                if(color_go){
+                    rgb_count++;
+                }
+                if(depth_go){
+                    depth_count++;
+                }
+                if(rectifL_go){
+                    left_count++;
+                }
+                if(rectifR_go){
+                    right_count++;
+                }
+        #endif
+
+        if(rectifR_go && rectifL_go && depth_go && color_go) {
+            #ifndef NDEBUG
+                all_count++;
+            #endif
+            auto colorFrame = color->tryGet<dai::ImgFrame>();
+            auto depthFrame = depthQueue->tryGet<dai::ImgFrame>();
+            auto rectifL = rectifLeftQueue->tryGet<dai::ImgFrame>();
+            auto rectifR = rectifRightQueue->tryGet<dai::ImgFrame>();
+            
+            cv::Mat color = cv::Mat(colorFrame->getHeight(), colorFrame->getWidth(), CV_8UC3, colorFrame->getData().data());
```

Comment:
Is this not equivalent to `.clone()`?

Reply:
Does not appear to be. 

<!--StartFragment-->
data | Pointer  to the user data. Matrix constructors that take data and step  parameters do not allocate matrix data. Instead, they just initialize  the matrix header that points to the specified data, which means that _no  data is copied._
-- | --


<!--EndFragment-->
[Doc Ref](https://docs.opencv.org/master/d3/d63/classcv_1_1Mat.html#a51615ebf17a64c968df0bf49b4de6a3a)